src/slicer.cpp

```diff
@@ -82,122 +84,305 @@ int SlicerLayer::getNextSegmentIdx(const Mesh* mesh, const SlicerSegment& segmen
+{
+    // below code closes smallest gaps first
+
+    struct PossibleStitch
     {
-        PolygonRef open_polyline = open_polylines[open_polyline_idx];
-        
-        if (open_polyline.size() < 1) continue;
-        for(unsigned int open_polyline_other_idx = 0; open_polyline_other_idx < open_polylines.size(); open_polyline_other_idx++)
+        int64_t dist2;
+        // polyline_idx*2 + {0 | front, 1 | back}
+        unsigned int terminus_0_idx;
+        unsigned int terminus_1_idx;
+
+        bool in_order() const
         {
-            PolygonRef open_polyline_other = open_polylines[open_polyline_other_idx];
-            
-            if (open_polyline_other.size() < 1) continue;
-            
-            Point diff = open_polyline.back() - open_polyline_other[0];
+            // in order if using back of line 1 and front of line 2
+            return (terminus_0_idx & 1) &&
+                !(terminus_1_idx & 1);
+        }
 
-            if (shorterThen(diff, largest_neglected_gap_second_phase))
+        // priority_queue will give greatest first so greatest
+        // must be most desirable stitch
+        bool operator<(const PossibleStitch &other) const
+        {
+            if (dist2 > other.dist2)
             {
-                if (open_polyline_idx == open_polyline_other_idx)
-                {
-                    polygons.add(open_polyline);
-                    open_polyline.clear();
-                    break;
-                }
-                else
-                {
-                    for (unsigned int line_idx = 0; line_idx < open_polyline_other.size(); line_idx++)
-                    {
-                        open_polyline.add(open_polyline_other[line_idx]);
-                    }
-                    open_polyline_other.clear();
-                }
+                return true;
+            }
+            else if (dist2 < other.dist2)
+            {
+                return false;
+            }
+
+            // use in order connections before reverse connections
+            if (!in_order() && other.in_order())
+            {
+                return true;
+            }
+
+            if (terminus_0_idx > other.terminus_0_idx)
+            {
+                return true;
+            }
+            else if (terminus_0_idx < other.terminus_0_idx)
+            {
+                return false;
+            }
+
+            if (terminus_1_idx > other.terminus_1_idx)
+            {
+                return true;
             }
+            else if (terminus_1_idx < other.terminus_1_idx)
+            {
+                return false;
+            }
+
+            return false;
         }
-    }
-}
+    };
+
+    std::priority_queue<PossibleStitch> stitch_queue;
 
-void SlicerLayer::stitch(Polygons& open_polylines)
-{ // TODO This is an inefficient implementation which can run in O(n^3) time.
-    // below code closes smallest gaps first
-    while(1)
     {
```

Comment:
It's in a separate block only because it is a separate logical step.  I think I need to make some private functions for the various steps.  I also need to add some comments.  Basically, this block is creating the set of possible stitches (in terms of the end points involved).